packages/amplify-category-api/src/provider-utils/awscloudformation/utils/containers-artifacts.ts

```diff
@@ -0,0 +1,327 @@
+    throw new Error('No definition available (docker-compose.yaml / docker-compose.yml / Dockerfile)');
+  }
+
+  if (containerDefinitionFiles[dockerComposeFileNameYaml] && containerDefinitionFiles[dockerComposeFileNameYml]) {
+    throw new Error('There should be only one docker-compose.yaml / docker-compose.yml)');
+  }
+
+  let composeContents = containerDefinitionFiles[dockerComposeFileNameYaml] || containerDefinitionFiles[dockerComposeFileNameYml];
+  const { [dockerfileFileName]: dockerfileContents } = containerDefinitionFiles;
+
+  const { buildspec, containers, service, secrets } = getContainers(composeContents, dockerfileContents);
+
+  const containersPorts = containers.reduce(
+    (acc, container) =>
+      [].concat(
+        acc,
+        container.portMappings.map(({ containerPort }) => containerPort),
+      ),
+    [],
+  );
+
+  let exposedContainer: { name: string; port: number };
+
+  const containersExposed = containers.filter(container => container.portMappings.length > 0);
+
+  if (containersPorts.length === 0) {
+    throw new Error('Service requires at least one exposed port');
+  } else if (containersPorts.length > 1) {
+    exposedContainer = await checkContainerExposed(containersExposed, exposedContainerFromMeta);
+  } else {
+    exposedContainer = {
+      name: containersExposed[0].name,
+      port: containersExposed[0].portMappings[0].containerPort,
+    };
+  }
+
+  fs.ensureDirSync(srcPath);
+  fs.writeFileSync(path.join(srcPath, 'buildspec.yml'), buildspec);
+
+  const secretsArns: Map<string, string> = new Map<string, string>();
+
+  if ((await shouldUpdateSecrets(context, secrets)) || isInitialDeploy) {
+    // Normalizes paths
+    // Validate secrets file paths, existence and prefixes
+    const errors = Object.entries(secrets).reduce((acc, [secretName, secretFilePath]) => {
+      const baseDir = path.isAbsolute(secretFilePath) ? '' : srcPath;
+      const normalizedFilePath = path.normalize(path.join(baseDir, secretFilePath));
+
+      secrets[secretName] = normalizedFilePath;
+
+      let canRead = true;
+
+      try {
+        const fd = fs.openSync(normalizedFilePath, 'r');
+
+        fs.closeSync(fd);
+      } catch (err) {
+        canRead = false;
+      }
+
+      if (!canRead) {
+        acc.push(`Secret file "${secretFilePath}" can't be read.`);
+        return acc;
+      }
+
+      const basename = path.basename(normalizedFilePath);
+      const hasCorrectPrefix = basename.startsWith('.secret-');
+
+      if (!hasCorrectPrefix) {
+        acc.push(`Secret file "${secretFilePath}" doesn't start with the ".secret-" prefix.`);
+        return acc;
+      }
+
+      const isInsideSrc = normalizedFilePath.startsWith(path.join(srcPath, path.sep));
+      if (isInsideSrc) {
+        acc.push(`Secret file "${secretFilePath}" should not be inside the "src" folder. The "src" folder will be uploaded to S3.`);
+        return acc;
+      }
+
+      return acc;
+    }, <string[]>[]);
```

Comment:
this reduce function is wayyy too long. needs to be broken out into at least a couple different functions